File: extensions/dragonbones/ArmatureCache.js
/****************************************************************************
 Copyright (c) 2018 Xiamen Yaji Software Co., Ltd.

 https://www.cocos.com/

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated engine source code (the "Software"), a limited,
 worldwide, royalty-free, non-assignable, revocable and non-exclusive license
 to use Cocos Creator solely to develop games on your target platforms. You shall
 not use Cocos Creator software for developing other software or tools that's
 used for developing games. You are not granted to publish, distribute,
 sublicense, and/or sell copies of Cocos Creator.

 The software or tools in this License Agreement are licensed, not sold.
 Xiamen Yaji Software Co., Ltd. reserves all rights not expressly granted to you.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 ****************************************************************************/
const MaxCacheTime = 30;
const FrameTime = 1 / 60; 

let _vertices = [];
let _indices = [];
let _vertexOffset = 0;
let _indexOffset = 0;
let _vfOffset = 0;
let _preTexUrl = null;
let _preBlendMode = null;
let _segVCount = 0;
let _segICount = 0;
let _segOffset = 0;
let _colorOffset = 0;
let _preColor = null;
let _x, _y;

//Cache all frames in an animation
let AnimationCache = cc.Class({
    ctor () {
        this.frames = [];
        this.totalTime = 0;
        this.isCompleted = false;
        this._frameIdx = -1;

        this._armatureInfo = null;
        this._animationName = null;
        this._tempSegments = null;
        this._tempColors = null;
    },

    init (armatureInfo, animationName) {
        this._armatureInfo = armatureInfo;
        this._animationName = animationName;
    },

    // Clear texture quote.
    clear () {
        for (let i = 0, n = this.frames.length; i < n; i++) {
            let frame = this.frames[i];
            frame.segments.length = 0;
        }
    },

    begin () {
        let armatureInfo = this._armatureInfo;
        if (armatureInfo.curAnimationCache) {
            armatureInfo.curAnimationCache.updateToFrame();
        }
        let armature = armatureInfo.armature;
        let animation = armature.animation;
        animation.play(this._animationName, 1);

        armatureInfo.curAnimationCache = this;
        this._frameIdx = -1;
        this.totalTime = 0;
        this.isCompleted = false;
    },

    end () {
        if (!this._needToUpdate()) {
            this._armatureInfo.curAnimationCache = null;
            this.frames.length = this._frameIdx + 1;
            this.isCompleted = true;
        }
    },

    _needToUpdate (toFrameIdx) {
        let armatureInfo = this._armatureInfo;
        let armature = armatureInfo.armature;
        let animation = armature.animation;
        return !animation.isCompleted && 
                this.totalTime < MaxCacheTime && 
                (toFrameIdx == undefined || this._frameIdx < toFrameIdx);
    },

    updateToFrame (toFrameIdx) {
        if (!this._needToUpdate(toFrameIdx)) return;

        let armatureInfo = this._armatureInfo;
        let armature = armatureInfo.armature;

        do {
            // Solid update frame rate 1/60.
            armature.advanceTime(FrameTime);
            this._frameIdx++;
            this._updateFrame(armature, this._frameIdx);
            this.totalTime += FrameTime;
        } while (this._needToUpdate(toFrameIdx));
       
        this.end();
    },

    updateAllFrame () {
        this.begin();
        this.updateToFrame();
    },

    _updateFrame (armature, index) {
        _vfOffset = 0;
        _indexOffset = 0;
        _vertexOffset = 0;
        _preTexUrl = null;
        _preBlendMode = null;
        _segVCount = 0;
        _segICount = 0;
        _segOffset = 0;
        _colorOffset = 0;
        _preColor = null;

        this.frames[index] = this.frames[index] || {
            segments : [],
            colors : [],
            vertices : null,
            uintVert : null,
            indices : null,
        };
        let frame = this.frames[index];

        let segments = this._tempSegments = frame.segments;
        let colors = this._tempColors = frame.colors;
        this._traverseArmature(armature);
        // At last must handle pre color and segment.
        // Because vertex count will right at the end.
        // Handle pre color.
        if (_colorOffset > 0) {
            colors[_colorOffset - 1].vfOffset = _vfOffset;
        }
        colors.length = _colorOffset;
        // Handle pre segment
        let preSegOffset = _segOffset - 1;
        if (preSegOffset >= 0) {
            if (_segICount > 0) {
                let preSegInfo = segments[preSegOffset];
                preSegInfo.indexCount = _segICount;
                preSegInfo.vfCount = _segVCount * 5;
                preSegInfo.vertexCount = _segVCount;
                segments.length = _segOffset;
            } else {
                segments.length = _segOffset - 1;
            }
        }

        // Discard all segments.
        if (segments.length === 0) return;

        // Fill vertices
        let vertices = frame.vertices || new Float32Array(_vfOffset);
        let uintVert = frame.uintVert || new Uint32Array(vertices.buffer);
        for (let i = 0, j = 0; i < _vfOffset;) {
            vertices[i++] = _vertices[j++]; // x
            vertices[i++] = _vertices[j++]; // y
            vertices[i++] = _vertices[j++]; // u
            vertices[i++] = _vertices[j++]; // v
            uintVert[i++] = _vertices[j++]; // color
        }

        // Fill indices
        let indices = frame.indices || new Uint16Array(_indexOffset);
        for (let i = 0; i < _indexOffset; i++) {
            indices[i] = _indices[i];
        }

        frame.vertices = vertices;
        frame.uintVert = uintVert;
        frame.indices = indices;
    },

    _traverseArmature (armature) {
        let colors = this._tempColors;
        let segments = this._tempSegments;
        let gVertices = _vertices;
        let gIndices = _indices;
        let slotVertices, slotIndices;
        let slots = armature._slots, slot, slotMatrix, slotMatrixm, slotColor, colorVal;
        let texture;
        let preSegOffset, preSegInfo;

        for (let i = 0, l = slots.length; i < l; i++) {
            slot = slots[i];
            if (!slot._visible || !slot._displayData) continue;

            slot.updateWorldMatrix();

            if (slot.childArmature) {
                this._traverseArmature(slot.childArmature);
                continue;
            }

            texture = slot.getTexture();
            if (!texture) continue;

            slotColor = slot._color;

            if (_preTexUrl !== texture.url || _preBlendMode !== slot._blendMode) {
                _preTexUrl = texture.url;
                _preBlendMode = slot._blendMode;
                // Handle pre segment.
                preSegOffset = _segOffset - 1;
                if (preSegOffset >= 0) {
                    if (_segICount > 0) {
                        preSegInfo = segments[preSegOffset];
                        preSegInfo.indexCount = _segICount;
                        preSegInfo.vertexCount = _segVCount;
                        preSegInfo.vfCount = _segVCount * 5;
                    } else {
                        // Discard pre segment.
                        _segOffset--;
                    }
                }
                // Handle now segment.
                segments[_segOffset] = {
                    tex : texture,
                    blendMode : slot._blendMode,
                    indexCount : 0,
                    vertexCount : 0,
                    vfCount : 0
                };
                _segOffset++;
                _segICount = 0;
                _segVCount = 0;
            }

            colorVal = ((slotColor.a<<24) >>> 0) + (slotColor.b<<16) + (slotColor.g<<8) + slotColor.r;

            if (_preColor !== colorVal) {
                _preColor = colorVal;
                if (_colorOffset > 0) {
                    colors[_colorOffset - 1].vfOffset = _vfOffset;
                }
                colors[_colorOffset++] = {
                    r : slotColor.r,
                    g : slotColor.g,
                    b : slotColor.b,
                    a : slotColor.a,
                    vfOffset : 0
                }
            }

            slotVertices = slot._localVertices;
            slotIndices = slot._indices;

            slotMatrix = slot._worldMatrix;
            slotMatrixm = slotMatrix.m;

            for (let j = 0, vl = slotVertices.length; j < vl;) {
                _x = slotVertices[j++];
                _y = slotVertices[j++];
                gVertices[_vfOffset++] = _x * slotMatrixm[0] + _y * slotMatrixm[4] + slotMatrixm[12];
                gVertices[_vfOffset++] = _x * slotMatrixm[1] + _y * slotMatrixm[5] + slotMatrixm[13];
                gVertices[_vfOffset++] = slotVertices[j++];
                gVertices[_vfOffset++] = slotVertices[j++];
                gVertices[_vfOffset++] = colorVal;
            }
            
            // This place must use segment vertex count to calculate vertex offset.
            // Assembler will calculate vertex offset again for different segment.
            for (let ii = 0, il = slotIndices.length; ii < il; ii ++) {
                gIndices[_indexOffset++] = _segVCount + slotIndices[ii];
            }

            _vertexOffset = _vfOffset / 5;
            _segICount += slotIndices.length;
            _segVCount += slotVertices.length / 4;
        }
    },
});

let ArmatureCache = cc.Class({
    ctor () {
        this._animationPool = {};
        this._armatureCache = {};
    },

    // If cache is private, cache will be destroy when dragonbones node destroy.
    dispose () {
        for (var key in this._armatureCache) {
            var armatureInfo = this._armatureCache[key];
            if (armatureInfo) {
                let armature = armatureInfo.armature;
                armature && armature.dispose();
            }
        }
        this._armatureCache = null;
        this._animationPool = null;
    },

    _removeArmature (armatureKey) {
        var armatureInfo = this._armatureCache[armatureKey];
        let animationsCache = armatureInfo.animationsCache;
        for (var aniKey in animationsCache) {
            // Clear cache texture, and put cache into pool.
            // No need to create TypedArray next time.
            let animationCache = animationsCache[aniKey];
            if (!animationCache) continue;
            this._animationPool[armatureKey + "#" + aniKey] = animationCache;
            animationCache.clear();
        }

        let armature = armatureInfo.armature;
        armature && armature.dispose();
        delete this._armatureCache[armatureKey];
    },

    // When db assets be destroy, remove armature from db cache.
    resetArmature (uuid) {
        for (var armatureKey in this._armatureCache) {
            if (armatureKey.indexOf(uuid) == -1) continue;
            this._removeArmature(armatureKey);
        }
    },

    getArmatureCache (armatureName, armatureKey, atlasUUID) {
        let armatureInfo = this._armatureCache[armatureKey];
        let armature;
        if (!armatureInfo) {
            let factory = dragonBones.CCFactory.getInstance();
            let proxy = factory.buildArmatureDisplay(armatureName, armatureKey, "", atlasUUID);
            if (!proxy || !proxy._armature) return;
            armature = proxy._armature;
            // If armature has child armature, can not be cache, because it's
            // animation data can not be precompute.
            if (!ArmatureCache.canCache(armature)) {
                armature.dispose();
                return;
            }

            this._armatureCache[armatureKey] = {
                armature : armature,
                // Cache all kinds of animation frame.
                // When armature is dispose, clear all animation cache.
                animationsCache : {},
                curAnimationCache: null,
            };
        } else {
            armature = armatureInfo.armature;
        }
        return armature;
    },

    getAnimationCache (armatureKey, animationName) {
        let armatureInfo = this._armatureCache[armatureKey];
        if (!armatureInfo) return null;

        let animationsCache = armatureInfo.animationsCache;
        return animationsCache[animationName];
    },

    initAnimationCache (armatureKey, animationName) {
        if (!animationName) return null;

        let armatureInfo = this._armatureCache[armatureKey];
        let armature = armatureInfo && armatureInfo.armature;
        if (!armature) return null;
        let animation = armature.animation;
        let hasAni = animation.hasAnimation(animationName);
        if (!hasAni) return null;

        let animationsCache = armatureInfo.animationsCache;
        let animationCache = animationsCache[animationName];
        if (!animationCache) {
            // If cache exist in pool, then just use it.
            let poolKey = armatureKey + "#" + animationName;
            animationCache = this._animationPool[poolKey];
            if (animationCache) {
                delete this._animationPool[poolKey];
            } else {
                animationCache = new AnimationCache();
            }
            animationCache.init(armatureInfo, animationName);
            animationsCache[animationName] = animationCache;
        }
        return animationCache;
    },

    updateAnimationCache (armatureKey, animationName) {
        let animationCache = this.initAnimationCache(armatureKey, animationName);
        if (!animationCache) return;
        animationCache.updateAllFrame();
        if (animationCache.totalTime >= MaxCacheTime) {
            cc.warn("Animation cache is overflow, maybe animation's frame is infinite, please change armature render mode to REALTIME, dragonbones uuid is [%s], animation name is [%s]", armatureKey, animationName);
        }
        return animationCache;
    }
});

ArmatureCache.FrameTime = FrameTime;
ArmatureCache.sharedCache = new ArmatureCache();
ArmatureCache.canCache = function (armature) {
    let slots = armature._slots;
    for (let i = 0, l = slots.length; i < l; i++) {
        let slot = slots[i];
        if (slot.childArmature) {
            return false;
        }
    }
    return true;
},

module.exports = ArmatureCache;

File: native/cocos/editor-support/dragonbones-creator-support/ArmatureCache.cpp
/**
 * The MIT License (MIT)
 *
 * Copyright (c) 2012-2020 DragonBones team and other contributors
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

#include "ArmatureCache.h"
#include "CCFactory.h"
#include "base/TypeDef.h"
#include "base/memory/Memory.h"

USING_NS_MW; // NOLINT(google-build-using-namespace)

using namespace cc; // NOLINT(google-build-using-namespace)

DRAGONBONES_NAMESPACE_BEGIN

float ArmatureCache::FrameTime = 1.0F / 60.0F;
float ArmatureCache::MaxCacheTime = 120.0F;

ArmatureCache::SegmentData::SegmentData() = default;

ArmatureCache::SegmentData::~SegmentData() {
    CC_SAFE_RELEASE_NULL(_texture);
}

void ArmatureCache::SegmentData::setTexture(cc::middleware::Texture2D *value) {
    CC_SAFE_ADD_REF(value);
    CC_SAFE_RELEASE(_texture);
    _texture = value;
}

cc::middleware::Texture2D *ArmatureCache::SegmentData::getTexture() const {
    return _texture;
}

ArmatureCache::FrameData::FrameData() = default;

ArmatureCache::FrameData::~FrameData() {
    for (auto &bone : _bones) {
        delete bone;
    }
    _bones.clear();

    for (auto &color : _colors) {
        delete color;
    }
    _colors.clear();

    for (auto &segment : _segments) {
        delete segment;
    }
    _segments.clear();
}

ArmatureCache::BoneData *ArmatureCache::FrameData::buildBoneData(std::size_t index) {
    if (index > _bones.size()) return nullptr;
    if (index == _bones.size()) {
        auto *boneData = new BoneData;
        _bones.push_back(boneData);
    }
    return _bones[index];
}

std::size_t ArmatureCache::FrameData::getBoneCount() const {
    return _bones.size();
}

ArmatureCache::ColorData *ArmatureCache::FrameData::buildColorData(std::size_t index) {
    if (index > _colors.size()) return nullptr;
    if (index == _colors.size()) {
        auto *colorData = new ColorData;
        _colors.push_back(colorData);
    }
    return _colors[index];
}

std::size_t ArmatureCache::FrameData::getColorCount() const {
    return _colors.size();
}

ArmatureCache::SegmentData *ArmatureCache::FrameData::buildSegmentData(std::size_t index) {
    if (index > _segments.size()) return nullptr;
    if (index == _segments.size()) {
        auto *segmentData = new SegmentData;
        _segments.push_back(segmentData);
    }
    return _segments[index];
}

std::size_t ArmatureCache::FrameData::getSegmentCount() const {
    return _segments.size();
}

ArmatureCache::AnimationData::AnimationData() = default;

ArmatureCache::AnimationData::~AnimationData() {
    reset();
}

void ArmatureCache::AnimationData::reset() {
    for (auto &frame : _frames) {
        delete frame;
    }
    _frames.clear();
    _isComplete = false;
    _totalTime = 0.0F;
}

bool ArmatureCache::AnimationData::needUpdate(int toFrameIdx) const {
    return !_isComplete && _totalTime <= MaxCacheTime && (toFrameIdx == -1 || _frames.size() < toFrameIdx + 1);
}

ArmatureCache::FrameData *ArmatureCache::AnimationData::buildFrameData(std::size_t frameIdx) {
    if (frameIdx > _frames.size()) {
        return nullptr;
    }
    if (frameIdx == _frames.size()) {
        auto *frameData = new FrameData();
        _frames.push_back(frameData);
    }
    return _frames[frameIdx];
}

ArmatureCache::FrameData *ArmatureCache::AnimationData::getFrameData(std::size_t frameIdx) const {
    if (frameIdx >= _frames.size()) {
        return nullptr;
    }
    return _frames[frameIdx];
}

std::size_t ArmatureCache::AnimationData::getFrameCount() const {
    return _frames.size();
}

ArmatureCache::ArmatureCache(const std::string &armatureName, const std::string &armatureKey, const std::string &atlasUUID) {
    _armatureDisplay = dragonBones::CCFactory::getFactory()->buildArmatureDisplay(armatureName, armatureKey, "", atlasUUID);
    if (_armatureDisplay) {
        _armatureDisplay->addRef();
    }
}

ArmatureCache::~ArmatureCache() {
    if (_armatureDisplay) {
        _armatureDisplay->release();
        _armatureDisplay = nullptr;
    }

    for (auto &animationCache : _animationCaches) {
        delete animationCache.second;
    }
    _animationCaches.clear();
}

ArmatureCache::AnimationData *ArmatureCache::buildAnimationData(const std::string &animationName) {
    if (!_armatureDisplay) return nullptr;

    AnimationData *aniData = nullptr;
    auto it = _animationCaches.find(animationName);
    if (it == _animationCaches.end()) {
        auto *armature = _armatureDisplay->getArmature();
        auto *animation = armature->getAnimation();
        auto hasAni = animation->hasAnimation(animationName);
        if (!hasAni) return nullptr;

        aniData = new AnimationData();
        aniData->_animationName = animationName;
        _animationCaches[animationName] = aniData;
    } else {
        aniData = it->second;
    }
    return aniData;
}

ArmatureCache::AnimationData *ArmatureCache::getAnimationData(const std::string &animationName) {
    auto it = _animationCaches.find(animationName);
    if (it == _animationCaches.end()) {
        return nullptr;
    }
    return it->second;
}

void ArmatureCache::updateToFrame(const std::string &animationName, int toFrameIdx /*= -1*/) {
    auto it = _animationCaches.find(animationName);
    if (it == _animationCaches.end()) {
        return;
    }

    AnimationData *animationData = it->second;
    if (!animationData || !animationData->needUpdate(toFrameIdx)) {
        return;
    }

    if (_curAnimationName != animationName) {
        updateToFrame(_curAnimationName);
        _curAnimationName = animationName;
    }

    auto *armature = _armatureDisplay->getArmature();
    auto *animation = armature->getAnimation();

    // init animation
    if (animationData->getFrameCount() == 0) {
        animation->play(animationName, 1);
    }

    do {
        armature->advanceTime(FrameTime);
        renderAnimationFrame(animationData);
        animationData->_totalTime += FrameTime;
        if (animation->isCompleted()) {
            animationData->_isComplete = true;
        }
    } while (animationData->needUpdate(toFrameIdx));
}

void ArmatureCache::renderAnimationFrame(AnimationData *animationData) {
    std::size_t frameIndex = animationData->getFrameCount();
    _frameData = animationData->buildFrameData(frameIndex);

    _preColor = Color4B(0, 0, 0, 0);
    _color = Color4B(255, 255, 255, 255);

    _preBlendMode = -1;
    _preTextureIndex = -1;
    _curTextureIndex = -1;
    _preISegWritePos = -1;
    _curISegLen = 0;
    _curVSegLen = 0;
    _materialLen = 0;

    auto *armature = _armatureDisplay->getArmature();
    traverseArmature(armature);

    if (_preISegWritePos != -1) {
        SegmentData *preSegmentData = _frameData->buildSegmentData(_materialLen - 1);
        preSegmentData->indexCount = _curISegLen;
        preSegmentData->vertexFloatCount = _curVSegLen;
    }

    auto colorCount = _frameData->getColorCount();
    if (colorCount > 0) {
        ColorData *preColorData = _frameData->buildColorData(colorCount - 1);
        preColorData->vertexFloatOffset = static_cast<int>(_frameData->vb.getCurPos()) / sizeof(float);
    }

    _frameData = nullptr;
}

void ArmatureCache::traverseArmature(Armature *armature, float parentOpacity /*= 1.0f*/) {
    middleware::IOBuffer &vb = _frameData->vb;
    middleware::IOBuffer &ib = _frameData->ib;

    const auto &bones = armature->getBones();
    Bone *bone = nullptr;
    const auto &slots = armature->getSlots();
    CCSlot *slot = nullptr;
    // range [0.0, 1.0]
    Color4B preColor(0, 0, 0, 0);
    Color4B color;
    middleware::Texture2D *texture = nullptr;

    auto flush = [&]() {
        // fill pre segment count field
        if (_preISegWritePos != -1) {
            SegmentData *preSegmentData = _frameData->buildSegmentData(_materialLen - 1);
            preSegmentData->indexCount = _curISegLen;
            preSegmentData->vertexFloatCount = _curVSegLen;
        }

        SegmentData *segmentData = _frameData->buildSegmentData(_materialLen);
        segmentData->setTexture(texture);
        segmentData->blendMode = static_cast<int>(slot->_blendMode);

        // save new segment count pos field
        _preISegWritePos = static_cast<int>(ib.getCurPos() / sizeof(uint16_t));
        // reset pre blend mode to current
        _preBlendMode = static_cast<int>(slot->_blendMode);
        // reset pre texture index to current
        _preTextureIndex = _curTextureIndex;
        // reset index segmentation count
        _curISegLen = 0;
        // reset vertex segmentation count
        _curVSegLen = 0;
        // material length increased
        _materialLen++;
    };

    for (auto *i : bones) {
        bone = i;
        auto boneCount = _frameData->getBoneCount();
        BoneData *boneData = _frameData->buildBoneData(boneCount);
        auto &boneOriginMat = bone->globalTransformMatrix;
        auto &matm = boneData->globalTransformMatrix.m;
        matm[0] = boneOriginMat.a;
        matm[1] = boneOriginMat.b;
        matm[4] = -boneOriginMat.c;
        matm[5] = -boneOriginMat.d;
        matm[12] = boneOriginMat.tx;
        matm[13] = boneOriginMat.ty;
    }

    for (auto *i : slots) {
        slot = dynamic_cast<CCSlot *>(i); // TODO(zhakasi): refine the logic
        if (slot == nullptr) {
            return;
        }
        if (!slot->getVisible()) {
            continue;
        }
        slot->updateWorldMatrix();

        Mat4 *worldMatrix = &slot->worldMatrix;

        // If slots has child armature,will traverse child first.
        Armature *childArmature = slot->getChildArmature();
        if (childArmature != nullptr) {
            traverseArmature(childArmature, parentOpacity * static_cast<float>(slot->color.a) / 255.0F);
            continue;
        }

        texture = slot->getTexture();
        if (!texture) continue;
        _curTextureIndex = texture->getRealTextureIndex();

        auto vbSize = slot->triangles.vertCount * sizeof(middleware::V3F_T2F_C4B);
        vb.checkSpace(vbSize, true);

        // If texture or blendMode change,will change material.
        if (_preTextureIndex != _curTextureIndex || _preBlendMode != static_cast<int>(slot->_blendMode)) {
            flush();
        }

        // Calculation vertex color.
        color.a = static_cast<uint8_t>(slot->color.a * parentOpacity);
        color.r = static_cast<uint8_t>(slot->color.r);
        color.g = static_cast<uint8_t>(slot->color.g);
        color.b = static_cast<uint8_t>(slot->color.b);

        if (preColor != color) {
            preColor = color;
            auto colorCount = _frameData->getColorCount();
            if (colorCount > 0) {
                ColorData *preColorData = _frameData->buildColorData(colorCount - 1);
                preColorData->vertexFloatOffset = vb.getCurPos() / sizeof(float);
            }
            ColorData *colorData = _frameData->buildColorData(colorCount);
            colorData->color = color;
        }

        // Transform component matrix to global matrix
        middleware::Triangles &triangles = slot->triangles;
        middleware::V3F_T2F_C4B *worldTriangles = slot->worldVerts;

        for (int v = 0, w = 0, vn = triangles.vertCount; v < vn; ++v, w += 2) {
            middleware::V3F_T2F_C4B *vertex = triangles.verts + v;
            middleware::V3F_T2F_C4B *worldVertex = worldTriangles + v;

            vertex->vertex.z = 0; //reset for z value
            worldVertex->vertex.transformMat4(vertex->vertex, *worldMatrix);

            worldVertex->color.r = color.r;
            worldVertex->color.g = color.g;
            worldVertex->color.b = color.b;
            worldVertex->color.a = color.a;
        }

        vb.writeBytes(reinterpret_cast<char *>(worldTriangles), vbSize);

        auto ibSize = triangles.indexCount * sizeof(uint16_t);
        ib.checkSpace(ibSize, true);

        auto vertexOffset = _curVSegLen / VF_XYZUVC;
        for (int ii = 0, nn = triangles.indexCount; ii < nn; ii++) {
            ib.writeUint16(triangles.indices[ii] + vertexOffset);
        }

        _curISegLen += triangles.indexCount;
        _curVSegLen += static_cast<int32_t>(vbSize / sizeof(float));
    } // End slot traverse
}

void ArmatureCache::resetAllAnimationData() {
    for (auto &animationCache : _animationCaches) {
        animationCache.second->reset();
    }
}

void ArmatureCache::resetAnimationData(const std::string &animationName) {
    for (auto &animationCache : _animationCaches) {
        if (animationCache.second->_animationName == animationName) {
            animationCache.second->reset();
            break;
        }
    }
}

CCArmatureDisplay *ArmatureCache::getArmatureDisplay() {
    return _armatureDisplay;
}

DRAGONBONES_NAMESPACE_END
